src/app/AttributeAccessInterface.cpp

```diff
@@ -55,39 +55,82 @@ CHIP_ERROR AttributeReportBuilder::FinishAttribute(AttributeReportIBs::Builder &
-            mEncodeState.mCurrentEncodingListIndex = 0;
-        }
-        mCurrentEncodingListIndex = 0;
+        // Clear mAllowPartialData flag here since this encode procedure is not atomic.
+        // The most common error in this function is CHIP_ERROR_NO_MEMORY / CHIP_ERROR_BUFFER_TOO_SMALL, just revert and try
+        // next time is ok.
+        mEncodeState.mAllowPartialData = false;
+
+        AttributeReportBuilder builder;
+
+        mPath.mListOp = ConcreteDataAttributePath::ListOperation::ReplaceAll;
+        ReturnErrorOnFailure(builder.PrepareAttribute(mAttributeReportIBsBuilder, mPath, mDataVersion));
+
+        auto * attributeDataWriter = mAttributeReportIBsBuilder.GetAttributeReport().GetAttributeData().GetWriter();
+        TLV::TLVType outerType;
+        ReturnErrorOnFailure(
+            attributeDataWriter->StartContainer(TLV::ContextTag(AttributeDataIB::Tag::kData), TLV::kTLVType_Array, outerType));
+        VerifyOrDie(outerType == kAttributeDataIBType);
+
+        // Instead of reserving hardcoded amounts, we could checkpoint the
+        // writer, encode array end and FinishAttribute, check that this fits,
+        // measure how much the writer advanced, then restore the checkpoint,
+        // reserve the measured value, and save it.  But that's probably more
+        // cycles than just reserving this known constant.
+        ReturnErrorOnFailure(
+            mAttributeReportIBsBuilder.GetWriter()->ReserveBuffer(kEndOfAttributeReportIBByteCount + kEndOfListByteCount));
+
+        mEncodeState.mCurrentEncodingListIndex = 0;
     }
+    else
+    {
+        // For all elements in the list, a report with append operation will be generated. This will not be changed during encoding
+        // of each report since the users cannot access mPath.
+        mPath.mListOp = ConcreteDataAttributePath::ListOperation::AppendItem;
+    }
+
+    mCurrentEncodingListIndex = 0;
 
-    // After encoding the empty list, the remaining items are atomically encoded into the buffer. Tell report engine to not
+    // After encoding the initial list start, the remaining items are atomically encoded into the buffer. Tell report engine to not
     // revert partial data.
     mEncodeState.mAllowPartialData = true;
 
-    // For all elements in the list, a report with append operation will be generated. This will not be changed during encoding
-    // of each report since the users cannot access mPath.
-    mPath.mListOp = ConcreteDataAttributePath::ListOperation::AppendItem;
     return CHIP_NO_ERROR;
 }
 
+void AttributeValueEncoder::EnsureListEnded()
+{
+    if (!mEncodingInitialList)
+    {
+        // Nothing to do.
+        return;
+    }
+
+    // Unreserve the space we reserved just for this.  Crash if anything here
+    // fails, because that would mean that we've corrupted our data, and since
+    // mEncodeState.mAllowPartialData is true nothing will clean up for us here.
```

Comment:
How could this happen? Is it hypothetical?

Reply:
I can't think of a way anything under here can fail as things stand, which is why I was OK with the VerifyOrDie bits.  Mostly it's meant to catch someone changing the code in some way that causes bugs; it felt better than just silently ignoring errors here.  @tcarmelveilleux